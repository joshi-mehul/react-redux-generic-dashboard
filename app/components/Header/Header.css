@import '../../styles/variables.css';

:root {
  --navigationHeight: 84px;
  --mobileNavigationHeight: 54px;
}

.root {
  background: #222;
}

.header {
  display: flex;
  align-items: center;
  padding: 0 var(--rootPadding);
  max-width: var(--containerMaxWidth);
  margin: 0 auto;
}

.navigationItems {
  flex: 1;
  display: flex;
  list-style: none;
  flex-direction: row;
  justify-content: flex-end;
  margin: 0;

  @media (--small-viewport) {
    flex-direction: column;
    transform: scaleY(0);
    transform-origin: 0 0;
    transition: all 0.3s;
    position: absolute;
    opacity: 0.5;
    left: 0;
    right: 0;
    top: var(--mobileNavigationHeight);
    max-height: 0;
    background: #333;

    &.openMenu {
      transform: scaleY(1);
      max-height: 1000px;
      opacity: 1;
    }
  }
}

.navigationItems li {
  flex: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  height: var(--navigationHeight);
  line-height: var(--navigationHeight);

  @media (--small-viewport) {
    flex: 1;
    height: var(--mobileNavigationHeight);
    line-height: var(--mobileNavigationHeight);
  }

  a {
    color: #ccc;
    text-decoration: none;
    width: 100%;
    text-align: center;
    padding: 0 20px;
    text-transform: uppercase;
    font-weight: 700;
    letter-spacing: 1px;

    &:hover,
    &.activeItem {
      text-decoration: underline;
      color: #ecf0f1;
    }
  }
}

.mobileMenu {
  display: none;
  height: var(--mobileNavigationHeight);
  line-height: var(--mobileNavigationHeight);
  padding: 0;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;

  @media (--small-viewport) {
    display: flex;
    flex: 1;
  }
}

.mobileTitle {
  color: #fff;
  font-weight: 700;
  font-size: 20px;
  text-transform: uppercase;
  text-decoration: none;
}

.desktopTitle {
  color: #fff;
  font-size: 32px;
  display: block;
  flex: 1;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: 700;

  @media (--small-viewport) {
    display: none;
  }
}

.searchButton {
  color: #fff;
  cursor: pointer;
  font-size: 32px;
  font-weight: 700;
  padding: 0 20px;
}

.backdrop {
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: fixed;
}

.modalContent {
  border: 0;
  background: transparent;
  outline: none;
}
