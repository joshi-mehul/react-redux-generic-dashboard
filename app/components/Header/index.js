// @flow

import styles from './Header.css';
import React, { Component } from 'react';
import { NavLink, Link } from 'react-router-dom';
import Modal from 'react-modal';
import cx from 'classnames';
import qs from 'query-string';
import HamburgerButton from 'app/components/HamburgerButton';
import Icon from 'app/components/Icon';
import Search from 'app/components/Search';

const ExactLink = props => <NavLink {...props} exact />;

const navigationItems = [
  {
    to: '/',
    label: 'Home',
    LinkComponent: ExactLink
  },
  {
    to: '/about',
    label: 'About',
    LinkComponent: NavLink
  },
  {
    to: '/contact',
    label: 'Contact',
    LinkComponent: NavLink
  }
];

type Props = {
  openSearch: () => void,
  closeSearch: () => void,
  searchOpen: boolean,
  location: {
    search: string
  }
};

type State = {
  menuOpen: boolean
};

/**
 * A responsive header component with titles and menus.
 * On small screens the menu collapses into an animated hamburger menu.
 */
class Header extends Component<Props, State> {
  state = {
    menuOpen: false
  };

  handleToggleMenu = () => {
    this.setState(state => ({ menuOpen: !state.menuOpen }));
  };

  handleCloseMenu = () => {
    this.setState({ menuOpen: false });
  };

  handleOpenSearch = () => {
    this.setState({ menuOpen: false });
    this.props.openSearch();
  };

  render() {
    return (
      <div className={styles.root}>
        <div className={styles.header}>
          <div className={styles.mobileMenu}>
            <HamburgerButton
              onClick={this.handleToggleMenu}
              open={this.state.menuOpen}
            />
            <Link to="/" className={styles.mobileTitle}>
              Mehul
            </Link>
          </div>

          <Link to="/" className={styles.desktopTitle}>
            Mehul
          </Link>
          <ul
            className={cx(
              styles.navigationItems,
              this.state.menuOpen && styles.openMenu
            )}
          >
            {navigationItems.map(({ LinkComponent, to, label }) => (
              <li key={label}>
                <LinkComponent
                  to={to}
                  activeClassName={styles.activeItem}
                  onClick={this.handleCloseMenu}
                >
                  {label}
                </LinkComponent>
              </li>
            ))}
          </ul>

          <button
            className={styles.searchButton}
            onClick={this.handleOpenSearch}
          >
            <Icon name="search" />
          </button>
        </div>

        <Modal
          isOpen={this.props.searchOpen}
          onRequestClose={this.props.closeSearch}
          className={styles.modalContent}
          overlayClassName={styles.backdrop}
          contentLabel="Search"
        >
          <Search
            query={qs.parse(this.props.location.search).q}
            onClose={this.props.closeSearch}
          />
        </Modal>
      </div>
    );
  }
}

export default Header;
