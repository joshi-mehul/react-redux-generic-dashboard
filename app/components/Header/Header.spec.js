import React from 'react';
import { MemoryRouter } from 'react-router';
import Header from './';
import renderer from 'react-test-renderer';

describe('<Header />', () => {
  it('should render correctly', () => {
    const component = renderer.create(
      <MemoryRouter>
        <Header location={{ query: {} }} />
      </MemoryRouter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
