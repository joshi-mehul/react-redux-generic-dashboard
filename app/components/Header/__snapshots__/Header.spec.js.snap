// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render correctly 1`] = `
<div
  className="root"
>
  <div
    className="header"
  >
    <div
      className="mobileMenu"
    >
      <div
        className="hamburgerButton"
        onClick={[Function]}
      >
        <div />
        <div />
        <div />
        <div />
      </div>
      <a
        className="mobileTitle"
        href="/"
        onClick={[Function]}
      >
        hanse
      </a>
    </div>
    <a
      className="desktopTitle"
      href="/"
      onClick={[Function]}
    >
      hanse
    </a>
    <ul
      className="navigationItems"
    >
      <li>
        <a
          aria-current="true"
          className="activeItem"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          Home
        </a>
      </li>
      <li>
        <a
          aria-current={false}
          className={undefined}
          href="/about"
          onClick={[Function]}
          style={undefined}
        >
          About
        </a>
      </li>
      <li>
        <a
          aria-current={false}
          className={undefined}
          href="/contact"
          onClick={[Function]}
          style={undefined}
        >
          Contact
        </a>
      </li>
    </ul>
    <button
      className="searchButton"
      onClick={[Function]}
    >
      <i
        className="icon ion-search"
      />
    </button>
  </div>
</div>
`;
